lightbug_http/header.mojo

```diff
@@ -1,885 +1,113 @@
-        self.__initialized = True
-    
-    fn next(inout self) raises -> Bool:
-        if not self.initialized():
-            self.set_next_colon(-1)
-            self.set_next_line(-1)
-            self.set_initialized()
-        
-        var b_len = len(self.__b)
-
-        if b_len >= 2 and (self.__b[0] == rChar_byte) and (self.__b[1] == nChar_byte):
-            self.set_b(self.__b[2:])
-            self.set_subslice_len(2)
-            return False
-        
-        if b_len >= 1 and (self.__b[0] == nChar_byte):
-            self.set_b(self.__b[1:])
-            self.set_subslice_len(self.subslice_len() + 1)
-            return False
-        
-        var colon: Int
-        if self.next_colon() >= 0:
-            colon = self.next_colon()
-            self.set_next_colon(-1)
-        else:
-            colon = index_byte(self.__b, colonChar_byte)
-            var newline = index_byte(self.__b, nChar_byte)
-            if newline < 0:
-                raise Error("Invalid header, did not find a newline at the end of the header")
-            if newline < colon:
-                raise Error("Invalid header, found a newline before the colon")
-        if colon < 0:
-            raise Error("Invalid header, did not find a colon")
-        
-        var jump_to = colon + 1
-        self.set_key(self.__b[:colon])
-
-        while len(self.__b) > jump_to and (self.__b[jump_to] == whitespace_byte):
-            jump_to += 1
-            self.set_next_line(self.next_line() - 1)
-        
-        self.set_subslice_len(self.subslice_len() + jump_to)
-        self.set_b(self.__b[jump_to:])
-
-        if self.next_line() >= 0:
-            jump_to = self.next_line()
-            self.set_next_line(-1)
-        else:
-            jump_to = index_byte(self.__b, nChar_byte)
-        if jump_to < 0:
-            raise Error("Invalid header, did not find a newline")
-        
-        jump_to += 1
-        self.set_value(self.__b[:jump_to-2]) # -2 to exclude the \r\n
-        self.set_subslice_len(self.subslice_len() + jump_to)
-        self.set_b(self.__b[jump_to:])
 
-        if jump_to > 0 and (self.value()[jump_to-1] == rChar_byte):
-            jump_to -= 1
-        while jump_to > 0 and (self.value()[jump_to-1] == whitespace_byte):
-            jump_to -= 1
-        self.set_value(self.value()[:jump_to])
-        
-        return True
-    
+        var first = r.read_word()
+        r.increment()
+        var second = r.read_word()
+        r.increment()
+        var third = r.read_line()
+
+        while not is_newline(r.peek()):
+            var key = r.read_until(BytesConstant.colon)
+            r.increment()
+            if is_space(r.peek()):
+                r.increment()
+            # TODO (bgreni): Handle possible trailing whitespace
+            var value = r.read_line()
+            self._inner[to_string(key^).lower()] = to_string(value^)
+
+        return (to_string(first^), to_string(second^), to_string(third^))
```

Comment:
Here I think I caught a bug when I uncommented `test_client.mojo` - it hangs indefinitely when trying to return this tuple. I'm not 100% sure but my first guess is it's related to a bug in Mojo I was experiencing before where returning tuples hang and the program never exits. 
We could file an issue in stdlib and for now replace this return with something else? What do you think

Reply:
There was an error in the creation of the HTTPRequest in that test which Seemingly caused httpbin to return nothing, which broke the parsing and caused an infinite loop, so still a bug that should be addressed at some point